TFNetworkRecLayer.py

```diff
@@ -5293,53 +5293,73 @@ def post_process_final_rec_vars_outputs(self, rec_vars_outputs, seq_len):
 
 class PositionalEncodingLayer(_ConcatInputLayer):
   """
-  Provides positional encoding in the form of (batch, time, n_out),
-  where n_out is the number of channels,
-  if it is run outside a :class:`RecLayer`,
-  or (batch, n_out)
+  Provides positional encoding in the form of (batch, time, n_out) or (time, batch, n_out) same as the source
+  where n_out is the number of channels, if it is run outside a :class:`RecLayer`,
+  and (batch, n_out) or (n_out, batch) same as the source
   if run inside a :class:`RecLayer`, where it will depend on the current time frame.
 
   Assumes one source input with a time dimension if outside a :class:`RecLayer`.
-  By default ("from" key not provided), it would either use "data", or ":i".
   With `add_to_input`, it will calculate `x + input`.
+  The output shape is the same as the input in all cases.
 
   The positional encoding is the same as in Tensor2Tensor.
   See :func:`TFUtil.get_positional_encoding`.
   """
   layer_class = "positional_encoding"
   recurrent = True
 
-  def __init__(self, add_to_input=False, constant=-1, **kwargs):
+  def __init__(self, add_to_input=False, constant=-1, offset=None, **kwargs):
     """
     :param bool add_to_input: will add the signal to the input
     :param int constant: if positive, always output the corresponding positional encoding.
+    :param None|LayerBase offset: Specify the offset to be added to positions. Expect shape (batch, time) or (batch,).
     """
     super(PositionalEncodingLayer, self).__init__(**kwargs)
     assert len(self.sources) == 1, "%s: expect a single source" % self
     source = self.input_data
+    assert source.feature_dim_axis == source.batch_ndim - 1  # Must be last.
     if add_to_input:
       assert source.dim == self.output.dim
+      template_wo_feat = source.copy_template_excluding_axis(source.feature_dim_axis)
+    else:
+      template_wo_feat = self.output.copy_template_excluding_axis(self.output.feature_dim_axis)
```

Comment:
Why do you call it `template_wo_feat`? Better call it `output_templ_wo_feat` or so.
Also, you don't need the `else:`. The other code is actually wrong. In both cases, it should be `self.output.copy_template_excluding_axis(...)`.

Reply:
> Better call it output_wo_feat (or maybe output_templ_wo_feat) or so.

OK.

> Also, you don't need the else:..

I see.

> The other code is actually wrong.

I'm not sure what is wrong here. Isn't it same to use `source` or `output` if `add_to_input`?
Plus, the code with `source` was your suggestion :)

Reply:
Maybe "wrong" is the wrong word. What you want here is that `output_wo_feat` is always exactly as `self.output` (wo. feature axis).
Because of how `get_out_data_from_opts` is defined currently, this luckily is correct here, but would become incorrect if you change `get_out_data_from_opts`. In any case, it's confusing, as a reader of this code might not know how `get_out_data_from_opts` is defined, and also, should not need to know (the less a reader needs to know about other independent places in the code, the easier it is to understand some code).

Reply:
Now I see. Thank you for the explanation.

Reply:
Now I see. Thank you for the explanation.